File: internal/database/repositories/planner.go
package repositories

import (
	"errors"

	"github.com/ericp/chronos-bot-reminder/internal/database/models"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

// plannerItemRepository implementation
type plannerItemRepository struct {
	db *gorm.DB
}

// NewPlannerItemRepository creates a new planner item repository instance
func NewPlannerItemRepository(db *gorm.DB) PlannerItemRepository {
	return &plannerItemRepository{db: db}
}

// CountByAccountID returns how many planner items currently belong to an
// account, used to enforce the per-account item cap.
func (r *plannerItemRepository) CountByAccountID(accountID uuid.UUID) (int64, error) {
	var count int64
	err := r.db.Model(&models.PlannerItem{}).Where("account_id = ?", accountID).Count(&count).Error
	return count, err
}

func (r *plannerItemRepository) Create(item *models.PlannerItem) error {
	// Append at the end of its period by default
	if item.Position == 0 {
		var maxPosition int
		r.db.Model(&models.PlannerItem{}).
			Where("account_id = ? AND period = ?", item.AccountID, item.Period).
			Select("COALESCE(MAX(position), 0)").
			Scan(&maxPosition)
		item.Position = maxPosition + 1
	}
	return r.db.Create(item).Error
}

func (r *plannerItemRepository) GetByID(id uuid.UUID) (*models.PlannerItem, error) {
	var item models.PlannerItem
	err := r.db.First(&item, "id = ?", id).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return &item, nil
}

// GetByAccountID returns all of the account's planner items, ordered by
// period then position so the caller can render the two columns directly.
func (r *plannerItemRepository) GetByAccountID(accountID uuid.UUID) ([]models.PlannerItem, error) {
	var items []models.PlannerItem
	err := r.db.Where("account_id = ?", accountID).
		Order("period ASC, position ASC, created_at ASC").
		Find(&items).Error
	return items, err
}

// GetByDFMItemID returns all planner items linked to a given DFM item, used
// to keep the checked state in sync when the DFM item changes.
func (r *plannerItemRepository) GetByDFMItemID(dfmItemID uuid.UUID) ([]models.PlannerItem, error) {
	var items []models.PlannerItem
	err := r.db.Where("dfm_item_id = ?", dfmItemID).Find(&items).Error
	return items, err
}

func (r *plannerItemRepository) Update(item *models.PlannerItem) error {
	return r.db.Save(item).Error
}

func (r *plannerItemRepository) Delete(id uuid.UUID) error {
	return r.db.Delete(&models.PlannerItem{}, "id = ?", id).Error
}

// DeleteByAccountID removes every planner item for the account, used to
// "erase everything" and start a fresh plan.
func (r *plannerItemRepository) DeleteByAccountID(accountID uuid.UUID) error {
	return r.db.Delete(&models.PlannerItem{}, "account_id = ?", accountID).Error
}

// ReorderInput describes a single item's new position within the reorder transaction
type ReorderInput struct {
	ID       uuid.UUID
	Position int
	Period   models.PlannerPeriod
}

// Reorder atomically updates the position/period of multiple items, e.g.
// after a drag-and-drop reorder or a move between Morning/Afternoon.
func (r *plannerItemRepository) Reorder(accountID uuid.UUID, updates []ReorderInput) error {
	return r.db.Transaction(func(tx *gorm.DB) error {
		for _, u := range updates {
			res := tx.Model(&models.PlannerItem{}).
				Where("id = ? AND account_id = ?", u.ID, accountID).
				Updates(map[string]interface{}{
					"position": u.Position,
					"period":   u.Period,
				})
			if res.Error != nil {
				return res.Error
			}
		}
		return nil
	})
}

File: web/src/services/planner.ts
import { httpClient } from "./http";
import type { PlannerItem, PlannerPeriod } from "./types";

/**
 * Day Planner Service
 * Handles the account's day-planning checklist (web/mobile only feature,
 * no Discord bot surface). Items can optionally be linked to a DFM item so
 * the checked state stays in sync between the two.
 */
class PlannerService {
  /**
   * Fetch all of the account's planner items
   */
  async getItems(): Promise<PlannerItem[]> {
    try {
      const response = await httpClient.get<{ items: PlannerItem[] }>(
        "/api/planner/items"
      );
      return response.items || [];
    } catch (error) {
      console.error("Failed to fetch planner items:", error);
      return [];
    }
  }

  /**
   * Add a new item, optionally linked to an existing DFM item
   */
  async addItem(data: {
    content: string;
    period: PlannerPeriod;
    dfm_item_id?: string;
  }): Promise<PlannerItem> {
    try {
      return await httpClient.post<PlannerItem>("/api/planner/items", data);
    } catch (error) {
      console.error("Failed to add planner item:", error);
      throw error instanceof Error
        ? error
        : new Error("Failed to add planner item");
    }
  }

  /**
   * Update an item's content, checked state and/or period
   */
  async updateItem(
    itemId: string,
    data: { content?: string; checked?: boolean; period?: PlannerPeriod }
  ): Promise<PlannerItem | null> {
    try {
      return await httpClient.put<PlannerItem>(
        `/api/planner/items/${itemId}`,
        data
      );
    } catch (error) {
      console.error("Failed to update planner item:", error);
      return null;
    }
  }

  /**
   * Delete a single item
   */
  async deleteItem(itemId: string): Promise<boolean> {
    try {
      await httpClient.delete(`/api/planner/items/${itemId}`);
      return true;
    } catch (error) {
      console.error("Failed to delete planner item:", error);
      return false;
    }
  }

  /**
   * Persist a new order/period for multiple items at once, e.g. after a
   * drag-and-drop reorder or a move between Morning/Afternoon.
   */
  async reorder(
    items: Array<{ id: string; position: number; period: PlannerPeriod }>
  ): Promise<PlannerItem[]> {
    try {
      const response = await httpClient.put<{ items: PlannerItem[] }>(
        "/api/planner/reorder",
        { items }
      );
      return response.items || [];
    } catch (error) {
      console.error("Failed to reorder planner items:", error);
      throw error instanceof Error
        ? error
        : new Error("Failed to reorder planner items");
    }
  }

  /**
   * Erase every item and start a fresh plan
   */
  async clearAll(): Promise<boolean> {
    try {
      await httpClient.delete("/api/planner/items");
      return true;
    } catch (error) {
      console.error("Failed to clear planner items:", error);
      return false;
    }
  }
}

// Export singleton instance
export const plannerService = new PlannerService();
